kayak/SwiftFrame
# GH-34 Allow more flexible slicing

### Background

This PR changes the process of slicing the final canvas into slices to not rely on the size of the supplied screenshots anymore but rather the user specifying a desired number of slices.
It also fixes a couple of other things, like remove a bunch of dead code and QOL updates

### Changes

#### Changes relevant to #34

- added new key `numberOfSlices` to `DeviceData`
- added `SliceSizeCalculator` type and tests for it
- adapted the `validate()` method in `DeviceData` to verify a couple things
  - gap width is >= 0
  - number of slices is >= 1
  - resulting slice size is valid

#### Other changes

- updated `swift-argument-parser` to 1.3.0
- bumped minimum Swift version to 5.9 and minimum platform to macOS 13
- fixed CI by specifying which macOS version to use and also set the Xcode version before running/building tests
- updated example config files to contain the new required keys
- added `inversion: .prefixedNo` to two flags that had an inverted name, this makes it easier to reason about them without doing the mental inversion based on the name
- bumped version to 6.0.0 as the new slicing mechanism requires an update to the config files, which is therefore a breaking change
- moved regex parsing from `NSRegularExpression` to modern `Regex`
- removed old `clearDirectories` and `outputWholeImage` keys from parsing
- added tests for a bunch of things that were previously untested
- updated the `benchmark.py` script to not rely on any third-party libraries
- remove a lot of dead code and made many things internal that did not need to be public
  - I used [periphery](https://github.com/peripheryapp/periphery) to scan for such code

### How to Test

- run `swift run swiftframe render Example/example.yaml --verbose --manual-validation`
- check that the manual validation info contains the new key `Number of slices` and also that the `Output slice size` is displayed
- run tests

Closes #34
